### ISSUE TITLE
silent failure or hang when trying to run repl on multiple instances of emacs

### ISSUE THREAD

**Issue by tgbugs** [greghendershott/racket-mode#452]
I think I understand the cause of most of the behavior described below, but I'm not sure what to do about it. Note `C-c C-c`/`C-c C-k` mentioned below mean `racket-run-module-at-point `.

Reproduction.
1. run two instances of emacs.
2. run `M-x` `racket-repl` in the first
3. run `M-x` `racket-repl` in the second

Results.  
In the second instance of emacs the `*Racket REPL*` buffer shows only  
`Process *Racket REPL* connection broken by remote peer`.

`ps aux` or `htop` show two racket processes trying to run on the same port.
`/usr/bin/racket /home/tom/git/racket-mode/racket/main.rkt 55555 (auth rnd1)`
`/usr/bin/racket /home/tom/git/racket-mode/racket/main.rkt 55555 (auth rnd2)`

There are a number of issues related to this when trying to run the contents of a `racket-mode` buffer if a repl process already exists running in another instance of emacs, some of the error messages I have seen are included below. All of the behavior is quite confusing and there are usually no error messages of any kind to indicate what the issue is. In addition, the repl window never appears and I had to hunt it down and switch to it manually to see the connection broken message.

In all cases there is either a silent error (very confusing) or a very long hang (probably the duration of one of the timeouts, but it is more than 30 seconds, possibly more than 60 seconds if that helps pin down the possible issue).

Repro for the hang.
1. run a regular instance of emacs
2. run an instance of doom-emacs
3. run `M-x` `racket-repl` in the regular instance
4. run `C-c C-k` on a `racket-mode` buffer in the doom-emacs instance (`C-c C-c` also works)

Results.  
The doom emacs instance will freeze and one core will go to 100% cpu usage for a long time. When it is responsive again, the `*Racket REPL*` buffer will be empty and there will be nothing in `*Messages*` to indicate that anything has gone wrong.

Occasionally the following will appear in `*Messages*`. This happens when the first instance running on port 55555 is closed. I figured this out because I saw the `Process *Racket REPL* connection broken by remote peer` message appear in the second instance when I closed the first instance. Note that when this happens the second instance running on 55555 is still running according to `ps aux`/`htop`. Annoyingly there is no obvious way to kill the second racket process so that a new one can be started an connected to. `C-x C-k` on the second `*Racket REPL*` buffer closes it, but the error below still appears and the second racket process running on 55555 is not closed. Killing the second racket process manually and then trying to run the `racket-mode` buffer again with `C-c C-c` or `C-c C-k` causes the `Wrong type argument: hash-table-p, nil` error below.
*EDIT*: when I close the `racket-mode` buffer and say `yes` to closing the repl process, then running the buffer again with `C-c C-c` works as expected.

```
Error running timer: (file-error "make client process failed" "Connection refused"
:name "*Racket REPL*"
:buffer #<buffer *Racket REPL*>
:host "127.0.0.1"
:service 55555
:nowait nil
:tls-parameters nil)
```

From https://github.com/greghendershott/racket-mode/blob/202cc1b7841a5e24c1020d831c435cc953bf8e07/racket-repl.el#L638
```
compilation-forget-errors: Wrong type argument: hash-table-p, nil
```

<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20200420 1412)
                 "Major mode for Racket language."
                 ((emacs
                   (25 1))
                  (faceup
                   (0 0 2))
                  (pos-tip
                   (20191127 1028)))
                 nil nil "/home/tom/.emacs.d/elpa/racket-mode-20200420.1412"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott")))
                 nil))
((emacs-version "26.3")
 (emacs-uptime "2 minutes, 3 seconds")
 (system-type gnu/linux)
 (major-mode racket-repl-mode)
 (racket--el-source-dir "/home/tom/git/racket-mode/")
 (racket--rkt-source-dir "/home/tom/git/racket-mode/racket/")
 (racket-program "racket")
 (racket-command-port 55555)
 (racket-command-timeout 10)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-use-repl-submit-predicate nil)
 (racket-images-system-viewer "display")
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal))))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (completion-in-region-mode)
 (csv-field-index-mode)
 (diff-auto-refine-mode)
 (display-line-numbers-mode)
 (electric-indent-mode)
 (electric-pair-mode)
 (evil-local-mode)
 (evil-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-display-line-numbers-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-magit-file-mode)
 (global-undo-tree-mode)
 (line-number-mode)
 (magit-auto-revert-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (override-global-mode)
 (rainbow-delimiters-mode)
 (savehist-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (undo-tree-mode)
 (url-handler-mode)
 (which-function-mode))
(disabled-minor-modes
 (abbrev-mode)
 (archive-subfile-mode)
 (auto-complete-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (bibtex-completion-notes-global-mode)
 (bibtex-completion-notes-mode)
 (blink-cursor-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (bug-reference-mode)
 (bug-reference-prog-mode)
 (cider--debug-mode)
 (cider-auto-test-mode)
 (cider-enlighten-mode)
 (cider-mode)
 (cider-popup-buffer-mode)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (csv-align-mode)
 (defining-kbd-macro)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (display-time-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (elpy-django)
 (elpy-mode)
 (eval-sexp-fu-flash-mode)
 (evil-leader-mode)
 (evil-magit-toggle-text-minor-mode)
 (evil-org-mode)
 (evil-paredit-mode)
 (flycheck-mode)
 (flyspell-mode)
 (git-commit-mode)
 (global-auto-complete-mode)
 (global-auto-revert-mode)
 (global-evil-leader-mode)
 (global-flycheck-mode)
 (global-hl-line-mode)
 (global-hl-todo-mode)
 (global-linum-mode)
 (global-prettify-symbols-mode)
 (global-reveal-mode)
 (global-visual-line-mode)
 (gnuplot-context-sensitive-mode)
 (goto-address-mode)
 (goto-address-prog-mode)
 (helm--minor-mode)
 (helm--remap-mouse-mode)
 (helm-autoresize-mode)
 (helm-display-line-numbers-mode)
 (helm-migemo-mode)
 (helm-popup-tip-mode)
 (highlight-numbers-mode)
 (hl-line-mode)
 (hl-todo-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (html-autoview-mode)
 (ido-everywhere)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (js2-highlight-unused-variables-mode)
 (js2-minor-mode)
 (linum-mode)
 (macrostep-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-file-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (markdown-live-preview-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (ob-ipython-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (pycoverage-mode)
 (pyvenv-mode)
 (pyvenv-tracking-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (rainbow-delimiters-org-mode)
 (recentf-mode)
 (rectangle-mark-mode)
 (reftex-mode)
 (reveal-mode)
 (server-mode)
 (sgml-electric-tag-pair-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (slime-autodoc-mode)
 (slime-edit-value-mode)
 (slime-editing-mode)
 (slime-fuzzy-target-buffer-completions-mode)
 (slime-macroexpansion-minor-mode)
 (slime-mode)
 (slime-popup-buffer-mode)
 (slime-repl-map-mode)
 (slime-repl-read-mode)
 (slime-trace-dialog-autofollow-mode)
 (slime-trace-dialog-hide-details-mode)
 (slime-trace-dialog-minor-mode)
 (smerge-mode)
 (table-fixed-width-mode)
 (table-mode-indicator)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (transient-resume-mode)
 (undo-tree-visualizer-selection-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (vdiff-3way-mode)
 (vdiff-mode)
 (vdiff-scroll-lock-mode)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode)
 (with-editor-mode)
 (xref-etags-mode))
```
</details>

**Comment by greghendershott** [comment]
Thanks for the report. I need to digest, but meanwhile a few quick thoughts:

- Multiple Emacs running multiple Racket Mode back ends, is a use case I haven't really considered or tried to support (yet).

- You _can_ explicitly set `racket-command-port` to some other value for the second instance. (e.g. I do this when running tests interactively -- starting a second back end from the _same_ Emacs.) Just be sure to set it at least +2 from the other value, because the logger uses +1.  So if you have 55555 in one instance, use 55557 in the other. (That's what the tests do.)

- If you get into this situation, try `M-x racket-stop-back-end` to kill the extra one. Does that help?

Again these are just a few quick ideas.

---

Instead maybe the back end should pick a free port, itself, and tell the front end what port to use. Or something like that.

**Comment by greghendershott** [comment]
Thinking about this a bit more. Here's my plan:

1. Change the logger so that it doesn't need an extra TCP port. It could send log message as "notifications" over the main command I/O -- much like debug-break notifications work, now. (I've wanted to change this for awhile, but this provides extra motivation.)

2. Change the REPL server to request an "ephemeral" port -- pass `0` as the port to `tcp-listen`, and use `tcp-addresses` to learn the port chosen -- and return this to the front end along with the authorization cookie. As a result, `racket-command-port` would become obsolete. (It had already become misnamed. Ever since the recent "big switcheroo", the command I/O is now over stdin/stdout to the back end process. The TCP port is used for to create one or more REPL sessions and I/O for those.)

Doing 1 would make 2 simpler, as just one port per back end would be involved.

I think this would make things simpler, as well as support N of Emacs + Racket Mode back end combinations, with no special effort by a user.

**Comment by tgbugs** [comment]
That sounds like it would simplify things immensely, and also help with another use case that I have which is that I often would like to run both `racket` and `racketcgc` at the same time and being able to spin up multiple interpreters would be much easier if the ports were chosen automatically.

I ran into the port + 1 issue as well and was confused by it, so moving to a single tcp port would prevent that confusion as well in the event that anyone else stumbles into it.

**Comment by greghendershott** [comment]
Update: I'm still working on this. I have an implementation, but I'm still testing.

At one point, using the ephemeral port, I got that same "connection refused" `file-error`. So even with that approach, _if_ for whatever reason the TCP listener is not created successfully, there is still an error-handling path I wanted to review. I _think_ I've addressed that, now, but want to sleep on it more. (Also I got side-tracked by Travis CI tests failing to download Racket installers, for a few days.)
